File: kirun-js/src/engine/model/ParameterReference.ts
import { Schema } from '../json/schema/Schema';
import { SchemaType } from '../json/schema/type/SchemaType';
import { TypeUtil } from '../json/schema/type/TypeUtil';
import { Namespaces } from '../namespaces/Namespaces';
import { isNullValue } from '../util/NullCheck';
import UUID from '../util/UUID';
import { ParameterReferenceType } from './ParameterReferenceType';

export class ParameterReference {
    private static readonly SCHEMA_NAME: string = 'ParameterReference';
    public static readonly SCHEMA: Schema = new Schema()
        .setNamespace(Namespaces.SYSTEM)
        .setName(ParameterReference.SCHEMA_NAME)
        .setType(TypeUtil.of(SchemaType.OBJECT))
        .setProperties(
            new Map([
                ['key', Schema.ofString('key')],
                ['value', Schema.ofAny('value')],
                ['expression', Schema.ofString('expression')],
                ['type', Schema.ofString('type').setEnums(['EXPRESSION', 'VALUE'])],
                ['order', Schema.ofInteger('order')],
            ]),
        );

    private key: string;
    private type: ParameterReferenceType;
    private value: any;
    private expression?: string;
    private order?: number;

    constructor(type: ParameterReferenceType | ParameterReference) {
        if (type instanceof ParameterReference) {
            let pv = type as ParameterReference;
            this.key = pv.key;
            this.type = pv.type;
            this.value = isNullValue(pv.value) ? undefined : JSON.parse(JSON.stringify(pv.value));
            this.expression = pv.expression;
            this.order = pv.order;
        } else {
            this.type = type as ParameterReferenceType;
            this.key = UUID();
        }
    }

    public getType(): ParameterReferenceType {
        return this.type;
    }
    public setType(type: ParameterReferenceType): ParameterReference {
        this.type = type;
        return this;
    }

    public getKey(): string {
        return this.key;
    }
    public setKey(key: string): ParameterReference {
        this.key = key;
        return this;
    }

    public getValue(): any {
        return this.value;
    }
    public setValue(value: any): ParameterReference {
        this.value = value;
        return this;
    }
    public getExpression(): string | undefined {
        return this.expression;
    }
    public setExpression(expression: string): ParameterReference {
        this.expression = expression;
        return this;
    }

    public setOrder(order: number): ParameterReference {
        this.order = order;
        return this;
    }
    public getOrder(): number | undefined {
        return this.order;
    }

    public static ofExpression(value: any): [string, ParameterReference] {
        const param = new ParameterReference(ParameterReferenceType.EXPRESSION).setExpression(
            value,
        );
        return [param.getKey(), param];
    }

    public static ofValue(value: any): [string, ParameterReference] {
        const param = new ParameterReference(ParameterReferenceType.VALUE).setValue(value);
        return [param.getKey(), param];
    }

    public static from(e: any): ParameterReference {
        return new ParameterReference(e.type)
            .setValue(e.value)
            .setExpression(e.expression)
            .setKey(e.key)
            .setOrder(e.order);
    }

    public toJSON(): any {
        return {
            key: this.key,
            type: this.type,
            value: this.value,
            expression: this.expression,
            order: this.order,
        };
    }
}

File: kirun-py/src/kirun_py/model/parameter_reference.py
from __future__ import annotations
import copy
from typing import Any, Optional, Tuple

from kirun_py.model.parameter_reference_type import ParameterReferenceType
from kirun_py.util.null_check import is_null_value
from kirun_py.util.uuid_ import uuid


class ParameterReference:
    def __init__(self, type_or_ref):
        if isinstance(type_or_ref, ParameterReference):
            pv = type_or_ref
            self._key = pv._key
            self._type = pv._type
            self._value = copy.deepcopy(pv._value) if not is_null_value(pv._value) else None
            self._expression = pv._expression
            self._order = pv._order
        else:
            self._type: ParameterReferenceType = type_or_ref
            self._key: str = uuid()
            self._value: Any = None
            self._expression: Optional[str] = None
            self._order: Optional[int] = None

    def get_type(self) -> ParameterReferenceType:
        return self._type

    def set_type(self, type_: ParameterReferenceType) -> ParameterReference:
        self._type = type_
        return self

    def get_key(self) -> str:
        return self._key

    def set_key(self, key: str) -> ParameterReference:
        self._key = key
        return self

    def get_value(self) -> Any:
        return self._value

    def set_value(self, value: Any) -> ParameterReference:
        self._value = value
        return self

    def get_expression(self) -> Optional[str]:
        return self._expression

    def set_expression(self, expression: str) -> ParameterReference:
        self._expression = expression
        return self

    def set_order(self, order: int) -> ParameterReference:
        self._order = order
        return self

    def get_order(self) -> Optional[int]:
        return self._order

    @staticmethod
    def of_expression(value: str) -> Tuple[str, ParameterReference]:
        param = ParameterReference(ParameterReferenceType.EXPRESSION).set_expression(value)
        return (param.get_key(), param)

    @staticmethod
    def of_value(value: Any) -> Tuple[str, ParameterReference]:
        param = ParameterReference(ParameterReferenceType.VALUE).set_value(value)
        return (param.get_key(), param)

    @staticmethod
    def from_value(e: dict) -> ParameterReference:
        return (
            ParameterReference(ParameterReferenceType(e.get('type', 'VALUE')))
            .set_value(e.get('value'))
            .set_expression(e.get('expression', ''))
            .set_key(e.get('key', uuid()))
            .set_order(e.get('order', 0))
        )

    def to_json(self) -> dict:
        return {
            'key': self._key,
            'type': self._type.value if self._type else None,
            'value': self._value,
            'expression': self._expression,
            'order': self._order,
        }
